include/nbl/video/IGPUObjectFromAssetConverter.h

```diff
@@ -591,76 +710,189 @@ auto IGPUObjectFromAssetConverter::create(const asset::ICPUMesh** const _begin,
-                    tmpView->regenerateMipMapLevels();
-                }
+                IGPUCommandBuffer::SImageMemoryBarrier imgbarrier;
+                // TODO layouts
+                imgbarrier.oldLayout = asset::EIL_UNDEFINED;
+                imgbarrier.newLayout = asset::EIL_UNDEFINED;
+                imgbarrier.srcQueueFamilyIndex = _params.transferQueue->getFamilyIndex();
+                imgbarrier.dstQueueFamilyIndex = _params.computeQueue->getFamilyIndex();
+                imgbarrier.image = gpuimg;
+                asset::IImage::SSubresourceRange subres;
+                subres.baseArrayLayer = 0u;
+                subres.layerCount = params.arrayLayers;
+                subres.baseMipLevel = 0u;
+                subres.levelCount = params.mipLevels;
+                imgbarrier.subresourceRange = subres;
+                asset::SMemoryBarrier membarrier;
+                membarrier.srcAccessMask = asset::EAF_TRANSFER_WRITE_BIT;
+                membarrier.dstAccessMask = asset::EAF_SHADER_READ_BIT;
+                imgbarrier.barrier = membarrier;
+
+                imgMemBarriers.push_back(imgbarrier);
             }
+
+            auto gpubuf_found = img2gpubuf.find(cpuimg);
+            auto& gpubuf = gpubuf_found->second;
+
+            // TODO layout
+            cmdbuf->copyBufferToImage(gpubuf.get(), gpuimg.get(), asset::EIL_UNDEFINED, count, regions.begin());
 		}
 
 		res->operator[](i) = std::move(gpuimg);
     }
 
+    if (!cmdbuf)
+        return res;
+
+    // for when mipmaps fill be generated with compute shader
+    cmdbuf->pipelineBarrier(asset::EPSF_TRANSFER_BIT, asset::EPSF_COMPUTE_SHADER_BIT, 0u, 0u, nullptr, 0u, nullptr, imgMemBarriers.size(), imgMemBarriers.data());
+    cmdbuf->end();
+
+    auto imgsFilledSemaphore = _params.device->createSemaphore();
+    auto* imgsFilledSemaphore_ptr = imgsFilledSemaphore.get();
+
+    IGPUQueue::SSubmitInfo submit;
+    submit.commandBufferCount = 1u;
+    auto* cmdbuf_ptr = cmdbuf.get();
+    submit.commandBuffers = &cmdbuf_ptr;
+    submit.signalSemaphoreCount = 1u;
+    submit.pSignalSemaphores = &imgsFilledSemaphore_ptr;
+    submit.waitSemaphoreCount = 0u;
+    submit.pWaitDstStageMask = nullptr;
+    submit.pWaitSemaphores = nullptr;
+    // dont event tell the queue to signal a fence after last submit if it'll never be touched by user
+    auto* signalFence = _params.transferFence ? fence.get() : nullptr;
+    _params.transferQueue->submit(1u, nullptr, signalFence);
+    if (_params.transferFence && fence)
+        _params.transferFence[0] = std::move(fence);
+    if (_params.transferSemaphore)
+        _params.transferSemaphore[0] = imgsFilledSemaphore;
+
+    if (!needToGenMips)
+        return res;
+
+    // TODO when mipmaps are generated with compute -- another submit for compute queue, wait on imgsFilledSemaphore
+    // now this code below instead:
+    // @matt How to synchronize mipmap generation and filling image storage??
```

Comment:
if the transfer and compute queue are the same, then signal an event after every image upload, and do a split barrier (signal after upload, await before mipmap) without ownership transfer.

if they are two separate queues, then you need the `imgsFilledSemaphore` (that the user never sees or interacts with) signalled in the transfer submit above, which shall be awaited by the submit of the compute command buffer.
In this case you need "force" a submit (end the outside commandbuffer, submit it, swap it).

Reply:
actually it would be good to "hybrid merge" the 3 loops together to allow the transfer and compute to overlap when many images are being uploaded.

```cpp

auto it = images.begin();
auto cmdUpload = [&it]() -> void
{
    auto& image = *(it++);
    auto& buffer = image.buffer;
    transferCmdBuf->upload(buffer);
    transferCmdBuf->copy(image,buffer);
};
auto cmdComputeMip = [&it]() -> void
{
    ...
    computeCmdbuf->bindPipeline();
    computeCmdbuf->bindDescriptorSets();
    computeCmdbuf->pushConstants();
    computeCmdbuf->dispatch();
};
Semaphore mipSem;
if (transferQueue!=computeQueue)
{
    constexpr uint32_t pipeliningDepthHeuristic = 8u;
    auto doUploadBatch = [&]() -> void
    {
        auto oldBegin = it;
        for (uint32_t i=0u; i<pipeliningDepthHeuristic&&it!=images.end(); i++)
            cmdUpload();
        transferCmdBuf->pipelineBarrier(ownership transfer of the uploaded images to computeQueue);
        transferCmdBuf->end();
        Semaphore transferSem = device->createSemaphore();
        transferQueue->submit(transferCmdBuf,nullptr,transferSem);
        addToZombieAndSwapAndBegin(core::smart_refctd_ptr(transferSem),&transferCmdBuf,transferCmdPool);
        //
        it = oldBegin;
        for (uint32_t i=0u; i<pipeliningDepthHeuristic&&it!=images.end(); i++,it++)
            cmdComputeMip();
        // optional: could refrain from ending and submitting the compute cmdbuffer for the very last batch
        computeCmdbuf->end();
        mipSem = device->semaphore();
        transferQueue->submit(computeCmdbuf,transferSem,mipSem);
        addToZombieAndSwapAndBegin(std::smart_refctd_ptr(mipSem),&computeCmdBuf,computeCmdPool);
    };
    for (; it!=images.end(); it++)
         doUploadBatch();
}
else
{
    auto queue = transferQueue;
    // need to use one cmdbuffer to do the job properly
    assert(transferCmdBuf==computeCmdBuf);
    auto cmdbuf = transferCmdBuf;
    for (; it!=images.end(); it++)
    {
        cmdUpload();
        transferCmdBuf->pipelineBarrier(a simple lightweight barrier to make sure the dispatch waits for the transfer to complete);
        cmdComputeMip();
    }
    // optional: could refrain from ending and submitting the cmdbuffer
    cmdbuf->end();
    mipSem = device->semaphore();
    queue->submit(cmdbuf,nullptr,mipSem);
    addToZombieAndSwapAndBegin(core::smart_refctd_ptr(mipSem),&cmdbuf,transferCmdPool);
}
```